src/main/java/lu/oop/server/app/services/UserService.java

```diff
@@ -90,4 +97,63 @@ public boolean saveUserWithTags(IUserModel user,
         userRepository.save((UserModel) user); // Save the user with the updated tags
         return true; // Successfully saved
     }
+
+    public List<IUserModel> getByUsername(String username){
+        List<IUserModel> users = userRepository.getListByLogin(username);
+        return users;
+    }
+
+    public List<IUserModel> getRecomendedUsers(Long id) {
+        Optional<UserModel> mbyUser = userRepository.findById(Long.valueOf(id));
+        UserModel currUser = mbyUser.get();
+        Set<TagModel> userTags = new HashSet<>(Arrays.asList(currUser.getTags()));
+
+        List<UserModel> allUsers = userRepository.findAll();
+        List<IUserModel> filteredUsers = new LinkedList<IUserModel>();
+        List<IUserModel> priorityUsers = new LinkedList<IUserModel>();
+        
+        for(UserModel user : allUsers){
+            if (user.getId().equals(id)) {
+                continue; // Skip the current user
+            }
+
+            Set<TagModel> tags = new HashSet<>(Arrays.asList(user.getTags()));
+            tags.retainAll(userTags); //Common tags
+
+            boolean agesMatch = tags.stream().anyMatch(tag -> tag.getType().equals("age"));
+            boolean establishmentMatch = tags.stream().anyMatch(tag -> tag.getType().equals("establishment"));
+
+            //If age group doesnt match dont show, show higher if establisment matches
+            if (agesMatch) {
+                if(establishmentMatch){
+                    priorityUsers.add(user);
+                }
+                else{
+                    filteredUsers.add(user);
+                }
+            }
+        }
+
+        priorityUsers.addAll(filteredUsers);
+
+        return priorityUsers;
+    }
+
+    public List<ITeacherModel> getRecomendedTeachers(Long id) {
+        Optional<UserModel> mbyUser = userRepository.findById(Long.valueOf(id));
+        UserModel currUser = mbyUser.get();
+        Set<TagModel> userTags = new HashSet<>(Arrays.asList(currUser.getTags()));
+
+        List<ITeacherModel> allTeachers = userRepository.getAllTeachers();
+        List<ITeacherModel> filteredTeachers = new LinkedList<ITeacherModel>();
+
+        for(ITeacherModel teacher: allTeachers){
+            String teacherSubject = teacher.getSubject();
```

Comment:
This matches not only the interest tags, but also ages and establishment tags. Do not think this is the desired behavior.